### ISSUE TITLE
BH: tune csr settings for branch prediction history

### ISSUE THREAD

**Issue by pgkeller** [tenstorrent/tt-metal#15386]
```
      logic [1:0]   cfgBpHistSel;           // 27:26- BP history settings (no history/2-bit/4-bit/8-bit)
```

The above bit in the csr is set to 0 by default which disables branch prediction history.
This was done by HW by default because enabling history slows down dhrystone.
Speculation as to why that is the case is because the history table is too small which wouldn't work well for running large programs.
However, metal kernels are small.  Enabling the history could be a win.
Per HW, it may also be something that would best be set on a per-kernel basis for a small win

Bit can be set with:
```
 asm(R"ASM(
        .option push
        lui   t1, 0x<value>
        csrrs zero, 0x7c0, t1
        .option pop
         )ASM"
```

@abhullar-tt @arikTT @davorchap

**Comment by abhullar-tt** [comment]
Would this be part of a launch msg that worker receives depending on the size of the kernel that needs to be run. The worker would need to disable it before it sends the done signal to dispatcher if its per-kernel basis. 

Before adding this it would probably be good to test this through the pgm sweep test to measure any gains.

**Comment by pgkeller** [comment]
> Would this be part of a launch msg that worker receives depending on the size of the kernel that needs to be run. The worker would need to disable it before it sends the done signal to dispatcher if its per-kernel basis.
> 
> Before adding this it would probably be good to test this through the pgm sweep test to measure any gains.

Yes first step is to run a bunch of tests, both w/ dispatch and even w/ key ops/models.  Ideally, we'll find a setting that is optimal and leave it there, but if not we'll have to evaluate different options.  Could be meta data passed from the host or could be the above code wrapped into an API executed in the actual kernel.  Definitely prefer just a single setting, anything exposed would have to be a significant win.

**Comment by ruizhangTT** [comment]
I ran sweep_pgm_dispatch.sh on BH with different cfgBpHistSel settings.  [Results](https://docs.google.com/spreadsheets/d/16ViSG61C2VozpSOFXGbbJImyCUrl6ckR5e5FEjOFUpk/edit?gid=0#gid=0)

For 3 test scenarios, the run time is consistently longer.  For other test scenarios, I don't see meaningful difference.

**Comment by ruizhangTT** [comment]
Saw one failure on CI with cfgBpHistSel=1: https://github.com/tenstorrent/tt-metal/actions/runs/16818417795/job/47640540471

Same failure on CI with [cfgBpHistSel=2](https://github.com/tenstorrent/tt-metal/actions/runs/16897518094), [3](https://github.com/tenstorrent/tt-metal/actions/runs/16899728081).

Edit: failure is related to #21875

**Comment by ruizhangTT** [comment]
I manually ran the yolov11 model test: `pytest models/demos/yolov11/tests/perf -m models_device_performance_bare_metal` on bh-30:

| cfgBpHistSel | AVG DEVICE KERNEL SAMPLES/S |
|--------|--------|
| 3 | 273.79969635613674 |
| 2 | 275.1706883780009 |
| 1 | 275.76781342873426 |
| 0 | 279.261743654475 | 

Branch predictor history does not have any performance gain here.

**Comment by ruizhangTT** [comment]
tt-fabric test depends on #26733

**Comment by pgkeller** [comment]
closing this issue, current setting is best
